modules/dmrpp_module/build_dmrpp_util.cc

```diff
@@ -0,0 +1,963 @@
+// -*- mode: c++; c-basic-offset:4 -*-
+
+// This file is part of the Hyrax data server.
+
+// Copyright (c) 2022 OPeNDAP, Inc.
+// Author: James Gallagher <jgallagher@opendap.org>
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+//
+// You can contact OPeNDAP, Inc. at PO Box 112, Saunderstown, RI. 02874-0112.
+
+#include <iostream>
+#include <sstream>
+#include <memory>
+#include <iterator>
+
+#include <cstdlib>
+
+#include <H5Ppublic.h>
+#include <H5Dpublic.h>
+#include <H5Epublic.h>
+#include <H5Zpublic.h>  // Constants for compression filters
+#include <H5Spublic.h>
+#include <H5Tpublic.h>
+
+#include "h5common.h"   // This is in the hdf5 handler
+
+#include <libdap/Array.h>
+#include <libdap/util.h>
+#include <libdap/D4Attributes.h>
+
+#include <BESDebug.h>
+#include <BESNotFoundError.h>
+#include <BESInternalError.h>
+
+#include "DMRpp.h"
+#include "DmrppTypeFactory.h"
+#include "DmrppD4Group.h"
+
+#if 0
+#define H5S_MAX_RANK    32
+#define H5O_LAYOUT_NDIMS    (H5S_MAX_RANK+1)
+
+/*
+ * "Generic" chunk record.  Each chunk is keyed by the minimum logical
+ * N-dimensional coordinates and the datatype size of the chunk.
+ * The fastest-varying dimension is assumed to reference individual bytes of
+ * the array, so a 100-element 1-D array of 4-byte integers would really be a
+ * 2-D array with the slow varying dimension of size 100 and the fast varying
+ * dimension of size 4 (the storage dimensionality has very little to do with
+ * the real dimensionality).
+ *
+ * The chunk's file address, filter mask and size on disk are not key values.
+ */
+typedef struct H5D_chunk_rec_t {
+    hsize_t scaled[H5O_LAYOUT_NDIMS];    /* Logical offset to start */
+    uint32_t nbytes;                      /* Size of stored data */
+    uint32_t filter_mask;                 /* Excluded filters */
+    haddr_t chunk_addr;                  /* Address of chunk in file */
+} H5D_chunk_rec_t;
+#endif
+
+using namespace std;
+using namespace libdap;
+using namespace dmrpp;
+
+namespace build_dmrpp_util {
+
+bool verbose = false;   // Optionally set by build_dmrpp's main().
+
+#define VERBOSE(x) do { if (verbose) (x); } while(false)
+
+#if 0
+///////////////////////////////////////////////////////////////////////////////
```

Comment:
If you use this code, I think you should add a comment that they are adapted from the HDF5 handler . The same for the read_vlen_string().